File: dart/test/intermediario.dart
import 'package:dart/intermediario.dart';
import 'package:test/test.dart';

import 'shared.dart';

void main() {
  test('contar caracteres', () {
    var str = 'um macaco come banana';
    var strR = 5;
    expect(contarCaracteresA(str), strR, reason: formatarErro(str, strR.toString()));
  });

  test('contar caracteres repetidos', () {
    var str = 'Desta maneira, o fenômeno da Internet não pode mais se dissociar do processo de comunicação como um todo';
    var strR = false;
    expect(contarECompararCaracteresAO(str), strR, reason: formatarErro(str, strR.toString()));
  });

  test('index char', () {
    var str = 'Desta maneira, o fenômeno da Internet não pode mais se dissociar do processo de comunicação como um todo';
    var strR = [4,7,12,15,24,27,40,43,48,59,62,66,70,75,81,87,90,93,95,101,103];
    expect(obterIndexAO(str).join('-'), strR.join('-'), reason: formatarErro(str, strR.join(',')));
  });

  group('inversão de string', () {
    test('test 1', () {
      var str = 'um macaco come banana';
      var strR = 'um ocacam come ananab';
      expect(inverterParteDaString(str), strR, reason: formatarErro(str, strR));
    });
    test('test 2', () {
      var str = 'Desta maneira, o fenômeno da Internet não pode mais se dissociar do processo de comunicação como um todo';
      var strR = 'atseD ,arienam o onemônef da tenretnI não pode mais se raicossid do ossecorp de oãçacinumoc como um todo';
      expect(inverterParteDaString(str), strR, reason: formatarErro(str, strR));
    });
    test('test 3', () {
      var str = 'só mais esse teste para garantir';
      var strR = 'só mais esse etset para ritnarag';
      expect(inverterParteDaString(str), strR, reason: formatarErro(str, strR));
    });
  });

  group("mascarar string até os ultimos 4 digitos", (){
    test("test 1", () {
      var str = "123456780";
      var strR = "#####6780";
      expect(mascararString(str), strR, reason: formatarErro(str, strR));
    });
    test("trocar posicao de um item", () {
      var str = "1234 56780";
      var strR = "#### #6780";
      expect(mascararString(str), strR, reason: formatarErro(str, strR));
    });
  });

  test("trocar posicao de um item", () {
    var array = [1,2,3,4,5];
    var value = 2;
    var newPosition = 0;
    var strR = [2,1,3,4,5];
    expect(trocarPosicaoItem(array,value, newPosition).join("-"), strR.join("-"), reason: formatarErro(array.join(','), strR.join(",")));
  });

  test("reduzir digitos de uma string", () {
    var str = "942";
    var strR = 6;
    expect(reduzirNumerosDaString(str), strR, reason: formatarErro(str, strR.toString()));
  });
}

File: javascript/testes/intermediario.js
var assert = require("assert");
var shared = require("./shared.js");
var objTest = require("../codes/intermediario.js");

describe("contar caracteres a", function() {
    it("teste 1", function() {
        let str = "um macaco come banana";
        let strR = 5;
        assert.ok(objTest.contarCaracteresA(str) == strR, shared.formatarErro(str, strR.toString()));
    })
    it("teste 2", function() {
        let str = "um macaco come bananaaa";
        let strR = 7;
        assert.ok(objTest.contarCaracteresA(str) == strR, shared.formatarErro(str, strR.toString()));
    })
})

describe("contar caracteres repetidos a,o", function() {
    it("teste 1", function() {
        let str = "Desta maneira, o fenômeno da Internet não pode mais se dissociar do processo de comunicação como um todo";
        let strR = false;
        assert.ok(objTest.contarECompararCaracteresAO(str) == strR, shared.formatarErro(str, strR));
    })

    it("teste 2", function() {
        let str = "ao";
        let strR = true;
        assert.ok(objTest.contarECompararCaracteresAO(str) == strR, shared.formatarErro(str, strR));
    })
})

describe("index char a,o", function() {
    it("teste 1", function() {
        let str = "Desta maneira, o fenômeno da Internet não pode mais se dissociar do processo de comunicação como um todo";
        let strR = [4, 7, 12, 15, 24, 27, 40, 43, 48, 59, 62, 66, 70, 75, 81, 87, 90, 93, 95, 101, 103];
        assert.ok(objTest.obterIndexAO(str).join('-') == strR.join('-'), shared.formatarErro(str, strR.join(',')));
    })
    it("teste 2", function() {
        let str = "a";
        let strR = [0];
        assert.ok(objTest.obterIndexAO(str).join('-') == strR.join('-'), shared.formatarErro(str, strR.join(',')));
    })
})

describe("inversão de string", function() {
    it("teste 1", function() {
        let str = "um macaco come banana";
        let strR = "um ocacam come ananab";
        assert.ok(objTest.inverterParteDaString(str) == strR, shared.formatarErro(str, strR));
    })
    it("teste 2", function() {
        let str = "Desta maneira, o fenômeno da Internet não pode mais se dissociar do processo de comunicação como um todo";
        let strR = "atseD ,arienam o onemônef da tenretnI não pode mais se raicossid do ossecorp de oãçacinumoc como um todo";
        assert.ok(objTest.inverterParteDaString(str) == strR, shared.formatarErro(str, strR));
    })
    it("teste 3", function() {
        let str = "só mais esse teste para garantir";
        let strR = "só mais esse etset para ritnarag";
        assert.ok(objTest.inverterParteDaString(str) == strR, shared.formatarErro(str, strR));
    })
})

describe("mascarar string até os ultimos 4 digitos", function() {
    it("teste 1", function() {
        let str = "123456780";
        let strR = "#####6780";
        assert.ok(objTest.mascararString(str) == strR, shared.formatarErro(str, strR));
    })
    it("teste 2", function() {
        let str = "1234 56780";
        let strR = "#### #6780";
        assert.ok(objTest.mascararString(str) == strR, shared.formatarErro(str, strR));
    })
})

describe("trocar posição item array", function() {
    it("teste 1", function() {
        let array = [1,2,3,4,5];
        let value = 2;
        let newPosition = 0;
        let strR = [2,1,3,4,5];
        assert.ok(objTest.trocarPosicaoItem(array,value, newPosition).join("-") == strR.join("-"), shared.formatarErro(array.join(','), strR.join(",")));
    })

    it("teste 2", function() {
        let array = [1,2,3,4,5];
        let value = 5;
        let newPosition = 0;
        let strR = [5,1,2,3,4];
        assert.ok(objTest.trocarPosicaoItem(array,value, newPosition).join("-") == strR.join("-"), shared.formatarErro(array.join(','), strR.join(",")));
    })
})

describe("reduzir digitos de uma string", function() {
    it("teste 1", function() {
        let str = "942";
        let strR = 6;
        assert.ok(objTest.reduzirNumerosDaString(str) == strR, shared.formatarErro(str, strR));
    })
    it("teste 2", function() {
        let str = "16";
        let strR = 7;
        assert.ok(objTest.reduzirNumerosDaString(str) == strR, shared.formatarErro(str, strR));
    })
})